src/tools/transactions_builder.rs

```diff
@@ -0,0 +1,291 @@
+            )
+            .await?;
+            let out = runner.execute()?;
+            new_states.update_for_token(&cf.1, out.next_state.clone().try_into()?)?;
+            outputs.push(("hyllar".into(), out));
+            self.runners.push(runner);
+        }
+
+        for cf in self.staking_cf.iter() {
+            info!("State before runner: {:?}", new_states.staking);
+            info!("on chain state: {:?}", new_states.staking.on_chain_state());
+            let runner = ContractRunner::new::<staking::state::OnChainState>(
+                "staking".into(),
+                get_binary("staking".into())?,
+                self.identity.clone(),
+                BlobData(new_states.staking.as_digest().0),
+                self.blobs.clone(),
+                cf.1.clone(),
+                new_states.staking.on_chain_state(),
+            )
+            .await?;
+            let out = runner.execute()?;
+            let mut contract = StakingContract::new(
+                ExecutionContext {
+                    callees_blobs: vec![].into(),
+                    caller: self.identity.clone(),
+                },
+                new_states.staking.clone(),
+            );
+
+            contract
+                .execute_action(cf.0.clone())
+                .map_err(|e| anyhow!("Error in staking execution: {e}"))?;
+
+            new_states.staking = contract.state();
+            assert_eq!(
+                out.next_state,
+                new_states.staking.on_chain_state().as_digest()
+            );
+            outputs.push(("staking".into(), out));
+            self.runners.push(runner);
+        }
+
+        *states = new_states;
+
+        Ok(BuildResult {
+            identity: self.identity.clone(),
+            blobs: self.blobs.clone(),
+            outputs,
+        })
+    }
+
+    /// Returns an iterator over the proofs of the transactions
+    /// In order to send proofs when they are ready, without waiting for all of them to be ready
+    /// Example usage:
+    /// for (proof, contract_name) in transaction.iter_prove() {
+    ///    let proof: ProofData = proof.await.unwrap();
+    ///    ctx.client()
+    ///        .send_tx_proof(&hyle::model::ProofTransaction {
+    ///            blob_tx_hash: blob_tx_hash.clone(),
+    ///            proof,
+    ///            contract_name,
+    ///        })
+    ///        .await
+    ///        .unwrap();
+    ///}
+    pub fn iter_prove<'a>(
+        &'a self,
+    ) -> impl Iterator<
+        Item = (
+            Pin<Box<dyn std::future::Future<Output = Result<ProofData>> + Send + 'a>>,
+            ContractName,
+        ),
+    > + 'a {
+        self.runners.iter().map(|runner| {
+            let future = runner.prove();
+            (
+                Box::pin(future)
+                    as Pin<Box<dyn std::future::Future<Output = Result<ProofData>> + Send + 'a>>,
+                runner.contract_name.clone(),
+            )
+        })
+    }
+}
```

Comment:
Il te renvoie pas les preuves dans l'ordre de complétion si? Ou il y a de la magie que je ne suis pas. Tel que je lis le code là si on a 3 preuves et qu'elle finisse dans l'ordre on y gagne, mais sinon c'est équivalent à juste attendre les 3.

Tokio a des `JoinSet` qui exposent une API similaire: https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html

Reply:
cette fonction retourne juste un itérateur, donc si à chaque itération de la boucle tu .await sur la future, en effet tu gagnes rien de plus qu'attendre les 3. Tu gagnes juste que tu peux send les proof au fure et à mesure, plutôt que toutes d'un coup.

Mais, l'utilisateur de cette fonction peut tout à fait itérer là dessus et await toutes les futures en parallèle